### PR TITLE
Try finding commit using short SHA if it is not on HEAD

### PR DESCRIPTION
**Pull request by mikhainin**
# Pull Request Check List

If a dependency specifies a git revision using a short hash, Dulwich does not provide an option to directly address it (see discussion in #9560); however, we can still list revisions and find what we are looking for.

Resolves: #9560

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by mikhainin** [comment]
It seemed to work in my case, but I would appreciate it if someone looked into this and told me if it makes sense.
Happy to add tests - I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`?

**Comment by schneebuzz** [comment]
> I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`

Some tests for the `..backend.Git.clone` tests are in `tests/integration/test_utils_vcs_git.py`.

**Comment by radoering** [comment]
@jelmer I'd appreciate if you could take a look if it makes sense (from a dulwich point of view).

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [inline src/poetry/vcs/git/backend.py:328]: What is this try for? If the object was found in the local repo, why would it be missing here?

Comment (mikhainin) [reply src/poetry/vcs/git/backend.py:328]: Hmm, this is a good question. It probably was me being too cautious. Let me remove it.

**Comment by radoering** [comment]
> Happy to add tests - I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`?

We have tests in [tests/integration/test_utils_vcs_git.py](https://github.com/python-poetry/poetry/blob/main/tests/integration/test_utils_vcs_git.py) and in [main/tests/vcs/git/test_backend.py](https://github.com/python-poetry/poetry/blob/main/tests/vcs/git/test_backend.py). Fast tests that do not require an external repository can be put into the latter.

> I can see some integration tests against https://github.com/python-poetry/test-fixture-vcs-repository but every commit there seems to have its own head/tag which does not allow to reproduce the initial issue...

I have no idea if it is feasible, but maybe we can create a local repository in the test?

**Comment by jelmer** [comment]
Creating a local repository in e.g. a tempdir should be fast and easy with Dulwich - Repo.init(path) will do what you need.

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [reply src/poetry/vcs/git/backend.py:328]: You've retrieved the commit object, so that will definitely be present. 

One of git's basic invariants (ensured by e.g. fetch, etc) is that if you have the commit object you also have the other objects it refers to*.

[*] This isn't strictly true under some circumstances, e.g. shallows - but that's not relevant here.

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [inline src/poetry/vcs/git/backend.py:351]: With the upcoming release of Dulwich you can also use repo.object_store.iter_prefix, which avoids scanning the entire repository.

Comment (mikhainin) [reply src/poetry/vcs/git/backend.py:342]: Shall we wait till the next release and use the new method?

**Comment by mikhainin** [comment]
I added test, I had to introduce a new marker "skip_git_mock" not to mock "git clone" as that one is not ready to mock local clone. 

The rest was copied from `tests/vcs/git/test_system.py`: I moved fixtures into separate files so they can be re-used and added a "middle commit" so the issue was finally reproduced.

**Comment by jelmer** [comment]
> I added the test, I had to introduce a new marker "skip_git_mock" not to mock "git clone" as that one is not ready to mock local clone. 
> 
> The rest was copied from `tests/vcs/git/test_system.py`: I moved fixtures into separate files so they can be re-used and added a "middle commit" so the issue was finally reproduced.

LGTM, for what it's worth.

File: pyproject.toml
Comment (radoering) [inline pyproject.toml:196]:
```suggestion
    "skip_git_mock: mark tests that should not auto-apply git_mock"
```

File: tests/conftest.py
Comment (radoering) [inline tests/conftest.py:605]:
```suggestion
    # one commit which is not "head"
```

File: tests/conftest.py
Comment (radoering) [inline tests/conftest.py:616]: Is this an empty commit because it writes the same as the first commit? (It probably does not matter, I am just wondering.)

File: tests/vcs/git/test_backend.py
Comment (radoering) [inline tests/vcs/git/test_backend.py:87]: Do they? I thought they just use `dulwich` and do not need git installed (in comparison to `SystemGit`)?

**Review by radoering — CHANGES_REQUESTED** [review]

File: tests/conftest.py
Comment (mikhainin) [reply tests/conftest.py:616]: Thanks for flagging this, I added a new file instead

File: tests/vcs/git/test_backend.py
Comment (mikhainin) [reply tests/vcs/git/test_backend.py]: I thought the fixture uses `git cli` but now I see it's wasn't the case. Removed this `skipif`

File: tests/vcs/git/test_backend.py
Comment (radoering) [inline tests/vcs/git/test_backend.py:87]: I think we do not need this class anymore (without the `skipif`). There are already other tests for the `Git` class, which are top level. (In general, we tend to not create test classes but just test functions.)

File: tests/vcs/git/git_fixture.py
Comment (radoering) [inline tests/vcs/git/git_fixture.py:13]: I think we can put this constant back into `test_system` because it is unlikely that it is relevant for other tests than for `SystemGit`.

File: tests/conftest.py
Comment (radoering) [inline tests/conftest.py:590]: This fixture is only used in `test_backend` and `test_system`, isn't it? Then, we can put it in a `conftest.py` in `tests/vcs/git`.

**Review by radoering — CHANGES_REQUESTED** [review]

File: src/poetry/vcs/git/backend.py
Comment (radoering) [reply src/poetry/vcs/git/backend.py:342]: I do not think we have to wait unless @jelmer says there will be a release in the next days / few weeks. (It will take at least some weeks until the next Poetry release and I think it makes sense to include this with or without the improvement.)

File: tests/conftest.py
Comment (mikhainin) [reply tests/conftest.py:590]: Yes. However, it will interfere with Ruff's rules as ruff will be removed as an "unused import".
https://github.com/astral-sh/ruff/issues/3295
https://github.com/astral-sh/ruff/issues/8796

We can mark the import with `noqa` but I wasn't sure it was the best approach

Comment (radoering) [reply tests/conftest.py:590]: We do not have to import a fixture from a `conftest.py` to use it or do you mean something else?

File: tests/vcs/git/test_backend.py
Comment (mikhainin) [reply tests/vcs/git/test_backend.py]: removed the class

File: tests/vcs/git/git_fixture.py
Comment (mikhainin) [reply tests/vcs/git/git_fixture.py]: done

File: tests/conftest.py
Comment (mikhainin) [reply tests/conftest.py]: ah, sure, moved to `tests/vcs/git/conftest.py`

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [reply src/poetry/vcs/git/backend.py:342]: FWIW There's a new release out with this API now.

Comment (jelmer) [reply src/poetry/vcs/git/backend.py:342]:
```suggestion
        for sha in repo.object_store.iter_prefix(sha_prefix):
```

Comment (mikhainin) [reply src/poetry/vcs/git/backend.py:342]: I can see that this is implemented in **dulwich** `0.22.5`, which drops support of Python 3.8.
It is a reasonable change, but this pull-request branch still requires compatibility with `3.8`. I'm not sure it must be changed as a part of this PR...

@radoering?

Comment (Secrus) [reply src/poetry/vcs/git/backend.py:342]: We have dropped Python 3.8 on the main branch already, so that should not be an issue

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [inline src/poetry/vcs/git/backend.py:342]:
```suggestion
        for sha in repo.object_store.iter_prefix(prefix):
```

**Review by jelmer — APPROVED** [review]

File: src/poetry/vcs/git/backend.py
Comment (radoering) [reply src/poetry/vcs/git/backend.py:342]: @mikhainin As Secrus said. Just rebase your branch on main. That is required before merging anyway.

Comment (mikhainin) [reply src/poetry/vcs/git/backend.py:342]: Hmm, this direct replace causes:
```

~/src/poetry/src/poetry/vcs/git/backend.py:342: in _find_object_by_sha_prefix
    for sha in repo.object_store.iter_prefix(sha_prefix):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <DiskObjectStore('/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of-mikhailgalanin/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects')>
prefix = b'895d5f'

    def iter_prefix(self, prefix):
        if len(prefix) < 2:
            yield from super().iter_prefix(prefix)
            return
        seen = set()
        dir = prefix[:2].decode()
        rest = prefix[2:].decode()
>       for name in os.listdir(os.path.join(self.path, dir)):
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects/89'

.../virtualenvs/poetry-GxhiNo8B-py3.11/lib/python3.11/site-packages/dulwich/object_store.py:1042: FileNotFoundError



```

File: src/poetry/vcs/git/backend.py
Comment (abn) [inline src/poetry/vcs/git/backend.py:319]: @mikhainin can we move the new logic into an `else` block under the for loop here;

https://github.com/python-poetry/poetry/blob/6f6fd7012983a2e749c6030c1f5f155fd4397058/src/poetry/vcs/git/backend.py#L95-L101

And as @jelmer suggested, we can simplify it to something like 

```py
            else:
                for sha in repo.object_store.iter_prefix(prefix):
                    self.revision = sha.decode("utf-8")
                    break
```

Assuming we get a dulwich release with the fix soon.

**Review by abn — CHANGES_REQUESTED** [review]

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [reply src/poetry/vcs/git/backend.py:319]: I'll see if I can do a release this weekend,

**Comment by abn** [comment]
This will likely also require #9849 if we intend to update dulwich.

File: src/poetry/vcs/git/backend.py
Comment (jelmer) [reply src/poetry/vcs/git/backend.py:319]: 0.22.6 is now out with the fix

**Comment by dimbleby** [comment]
> This will likely also require #9849 if we intend to update dulwich.

Since recent dulwich releases are bumping only the patch number, poetry users are already getting the latest dulwich.

Are you saying that dulwich has released a breaking change in a patch release?

File: src/poetry/vcs/git/backend.py
Comment (dimbleby) [reply src/poetry/vcs/git/backend.py:319]: No it isn't!  I guess your release pipeline failed?

**Comment by jelmer** [comment]
I'm curious what change is, we haven't had reports of API breakages.

File: src/poetry/vcs/git/backend.py
Comment (mikhainin) [reply src/poetry/vcs/git/backend.py:319]: Right, I moved the logic to `_normalise()` - it now looks even cleaner.
Waiting for **0.22.6** to appear and then will push the change.

**Review by abn — APPROVED** [review]

File: src/poetry/vcs/git/backend.py
Comment (Secrus) [reply src/poetry/vcs/git/backend.py]: Dulwich 0.22.6 is out

**Comment by abn** [comment]
@jelmer To close the loop regarding @dimbleby 's question; I have raised a dulwich issue for triage with a reproducer.

Cross-post: https://github.com/jelmer/dulwich/issues/1443

**Review by radoering — APPROVED** [review]

**Pull request merged**
